activesupport/lib/active_support/cache.rb

```diff
@@ -843,49 +843,60 @@ def [](version)
       module Loader
         extend self
 
+        def dump(entry)
+          serialize(entry)
+        end
+
+        def dump_compressed(entry, threshold)
+          serialize(entry.compressed(threshold))
+        end
+
         def load(payload)
           if !payload.is_a?(String)
             ActiveSupport::Cache::Store.logger&.warn %{Payload wasn't a string, was #{payload.class.name} - couldn't unmarshal, so returning nil."}
 
             return nil
           elsif payload.start_with?(MARK_70_UNCOMPRESSED)
-            members = Marshal.load(payload.byteslice(1..-1))
+            members = deserialize(payload.byteslice(1..-1))
           elsif payload.start_with?(MARK_70_COMPRESSED)
-            members = Marshal.load(Zlib::Inflate.inflate(payload.byteslice(1..-1)))
+            members = deserialize(Zlib::Inflate.inflate(payload.byteslice(1..-1)))
           elsif payload.start_with?(MARK_61)
-            return Marshal.load(payload)
+            return deserialize(payload)
           else
             ActiveSupport::Cache::Store.logger&.warn %{Invalid cache prefix: #{payload.byteslice(0).inspect}, expected "\\x00" or "\\x01"}
 
             return nil
           end
           Entry.unpack(members)
         end
+
+        private
+          def deserialize(payload)
+            Marshal.load(payload)
+          rescue TypeError
```

Comment:
Here I was trying to add backward compatibility - if we have a JSON serialized values in our Cache and now we set our serializer to `:marshal`. But I'm not sure that this is the case, we need to support